In [1]:
from gapipy import Client
from datetime import timedelta
import pandas as pd


In [2]:
import re
from dotenv import load_dotenv
import os

# Load the environment variables from the .env file
load_dotenv()

# Access environment variables
api_key = os.getenv("G_ADVENTURES_KEY")

client = Client(application_key=api_key) 

In [3]:
tour_dossier = client.tour_dossiers.get(24309)

# Understanding a tour dossier

Let us this as an example

In [4]:
tour_dossier = client.tour_dossiers.get(24309)

In [5]:
tour_dossier

<TourDossier: 24309>

The tour dossier can be used as a dictionary with following keys:
- id: Tour ID
- href: Rest API for the tour
- name: Name of the tour
- slug: A version of the name field that is more suitable for use inside of a URL. Typically this means: all lowercase letters, hyphens instead of spaces, special characters and very common words removed. Sometimes the name field will change, but the slug should remain the same in order to preserve existing URL structures.
- product_line: A 1 to 10 character sequence for the |tour| associated with this product.
- departures_start_date: The earliest start_date of a departure for which this dossier is valid.
- departures_end_date: The latest start_date of a departure for which this dossier is valid.
- description: A short description of this tour dossier.
- details: A list of dictionaries describing the tour 
- categories: A list of dictionaries with all the categories of the trip inc Travel style, Service level and physical rating
- geography: A dictionary containing details of the geography of the tour
- images: A list of dictionaries containing image links including he map
- site_links: Links to the g-adventures site - needed for affiliate link

- __booking_companies__: List of companies you can book with
- __structured_itineraries__: A list of available itineraries
- __advertised_departures__: A list of gapipy.models
- __tour__: A reference to the tour
- __departures__: gapipy.query
- __relationships__: List of gapipy.resources

Example:

In [6]:
tour_dossier.description

"From the intricate saris of Rajasthan to the beautiful symmetry of the Taj Mahal, this compact journey is a photographer's delight and the perfect introduction to India. You'll float past the ghats on the Ganges River to encounter the spiritual heart of Hinduism in Varanasi and get off the beaten path exploring carved stepwells and quaint temples in rural villages. With the perfect combination of included activities and free time to explore the nation's rich heritage, you're sure to be caught up in India's incredible tapestry of life."

Those in bold are a little more complicated and have been broken down below

## booking_companies

A list of companies you can book with

In [7]:
tour_dossier.booking_companies

[<BookingCompany: 1 (stub)>]

In [8]:
tour_dossier.booking_companies[0]

<BookingCompany: 1 (stub)>

In [9]:
tour_dossier.booking_companies[0].to_dict()

{'id': '1', 'href': 'https://rest.gadventures.com/booking_companies/1'}

## structured_itineraries

structured_itineraries contains a list of itineraries available for a tour dossier. 

In [10]:
tour_dossier.structured_itineraries

[<Itinerary: 868 (stub)>, <Itinerary: 868 (stub)>]

An itinerary contains a structured representation of all activities, meals, transports, starting and finishing locations for a tour. 

In [11]:
tour_dossier.structured_itineraries[0].variation_id

'3543'

structured_itineraries can be used as a dictionary with following keys:
- id: the id of the tour dossier
- variation_id: The unique identifier of a specific variation of this itinerary
- href: Rest API for the itinerary can be accessed by - client.itineraries.get(868)
- name: A name to help disambiguate between itineraries on the same tour
- tour_dossier:A reference to the tour_dossier resource with which this itinerary is associated 
- booking_companies: The booking companies that provide this tour.
- flags: A list of potential flags for a itinerary
- duration: The length of the |tour| in days
- start_location: The starting location of |tour|
- end_location: The ending location of the |tour|
- meals_included: Information about the number of meals included in the |tour|
- meals_budget: Information about the budget for meals that are not included
- details: List of details for the itinerary
- packing_lists: List of dictionaries for packing list
- images: List of image objects
- media: gapipy.query
- highlights: gapipy.query
- valid_during_ranges: Dates the itinerary is valid for
- variations: The different variations of this itinerary
- __days__: A list on day objects accessed by .to_dict()
- site_links: List of dictionaries for the tour - useful for 
- publish_state: The state or online status of this itinerary
- ripple_score: The Ripple Score for an Itinerary. This is calculated based off the amount of money that stays locally. The score ranges between 0-100

Example:

In [12]:
tour_dossier.structured_itineraries[0].variation_id

'3543'

Days are a little more complicated and there is a lost of useful info so I've detailed it below

### Days

days is a list of days within the itinerary

In [13]:
tour_dossier.structured_itineraries[0].days

[<ItineraryDay 1>,
 <ItineraryDay 2>,
 <ItineraryDay 3>,
 <ItineraryDay 4>,
 <ItineraryDay 5>,
 <ItineraryDay 6>,
 <ItineraryDay 7>,
 <ItineraryDay 8>,
 <ItineraryDay 9>,
 <ItineraryDay 10>,
 <ItineraryDay 11>,
 <ItineraryDay 12>,
 <ItineraryDay 13>,
 <ItineraryDay 14>,
 <ItineraryDay 15>]

A day contains a structured representation of all activities, meals, transports, starting and finishing locations for a days.

In [14]:
tour_dossier.structured_itineraries[0].days[0]

<ItineraryDay 1>

days can be used as a dictionary with following keys:
- id: ID for the day
- day: The day number within the itinerary
- label: A location or activity based label for the day
- summary: 	A summary description of the day
- description: A detailed description of the day
- instructions: Additional notes, instructions or tips about the day needed when on the trip
- meals: Included meals for this day
- start_location: The starting location of the day
- end_location: The ending location of the day
- __components__: A list of components (activities and transports) for the day 
- optional_activities: A list of optional activities for the day

Example:

In [15]:
tour_dossier.structured_itineraries[0].days[8].summary

"Enjoy a day trip to Khajuraho, where India's largest group of medieval temples reside, known for their Kama Sutra statues. Learn the history of these world-famous temples on a guided tour and enjoy a taste of a little erotica from the Middle Ages. After, opt to visit a nearby waterfall or watch an evening dance performance."

#### components

components are a list of components within a day of the itinerary

In [16]:
tour_dossier.structured_itineraries[0].days[8].components

[<ItineraryComponent: TRANSPORT>,
 <ItineraryComponent: ACTIVITY>,
 <ItineraryComponent: TRANSPORT>,
 <ItineraryComponent: ACCOMMODATION>]

component contains a structured representation of the elements of a compontent.

In [17]:
tour_dossier.structured_itineraries[0].days[8].components[2]

<ItineraryComponent: TRANSPORT>

- type: TRANSPORT, ACTIVITY or ACCOMMODATION
- summary: A summary of the component
- description:
- instructions:
- distance_km:
- start_time:
- end_time:
- time_period:
- start_location: Link to the end location
- end_location: Link to the start location
- accommodation_dossier: The link to the dossier for the accommodation
- activity_dossier: The link to the dossier for the activity
- transport_dossier: The link to the dossier for the transport
- duration: min and max duration of the component if there is one

Useful example

In [18]:
itinerary = tour_dossier.structured_itineraries[1]
{day.day: day.summary for day in itinerary.days}

{1: 'Arrive at any time. Arrival transfer included through the G Adventures-supported Women on Wheels project.',
 2: 'Take a morning walk through the city with a young adult from the G Adventures-supported New Delhi Streetkids Project. Later, visit Old Delhi, explore the spice markets, and visit Jama Masjid and Connaught Place.',
 3: "Arrive in Jaipur and explore this gorgeous 'pink city'.",
 4: "Stop and take a photo of the 'Palace of the Winds' and wander the organized streets of the 'Pink City'. Later, enjoy a visit to the Amber Fort, famous for its mixture of Hindu and Muslim architecture.",
 5: 'Visit this Rajasthani rural village in the beautiful countryside with a close tie to the land.',
 6: 'Stop by the impressive Abhaneri stepwells on the way to Agra, home of the Taj Mahal.',
 7: 'Visit the famous Taj Mahal, Baby Taj, and Agra Fort.',
 8: 'Travel to Alipura and stay in a heritage property in the village.',
 9: "Enjoy a day trip to Khajuraho to visit the Khajuraho Group of Mon

## advertised_departures

advertised_departures contains a list of sellable departures, segmented by currency. Each object will contain the associated departure, pricing, currency, room, and a relevant promotion if applicable. This field is useful for showing the lowest price of a trip. 

In [19]:
tour_dossier.advertised_departures

In [20]:
tour_dossier.advertised_departures[7].to_dict()

{'previous_amount': None,
 'currency': 'JPY',
 'amount': '207000.00',
 'room': None,
 'departure': {'id': '1227042',
  'href': 'https://rest.gadventures.com/departures/1227042'},
 'promotion': None}

## tour

a reference to the tour, available through:

In [21]:
tour_dossier.tour.to_dict()

{'id': 24309, 'href': 'https://rest.gadventures.com/tours/24309'}

## departures

departures can be accessed individually

In [22]:
depart = client.departures.get(1080958)

or from the tour dossier

In [23]:
my_generator = tour_dossier.departures.all()

for value in my_generator:
    #print(value.sku, value.lowest_pp2a_prices[0].to_dict())
    break

- id: Departure ID
- href: Departure reference
- date_created: date the departure was created
- date_last_modified: date teh departure was last modified
- name: Tour name
- product_line: tour product line
- sku: refference number
- start_date: tour start date
- finish_date: tour end date
- date_cancelled: the date the tour was cancelled
- flags: A list of codes that, when present, require special considerations when booking a service on this departure.
- latest_arrival_time: The latest time that the traveller’s flight can arrive at the nearest_start_airport for this product, in the standard dates-time.
- earliest_departure_time: The earliest time that the traveller’s flight can depart at the nearest_finish_aiport for this product, in the standard dates-time.
- nearest_start_airport: The nearest airport to the start location
- nearest_finish_airport:The nearest airport to the end location
- availability: Availability information for this option 
- requirements: A list of requirements that must be met before a service on this product can be booked or confirmed. 
- program: Generally referencing another tour code
- booking_companies: A list of booking companies that own this product
- start_address: Start Location information for the product.
- finish_address: End Location information for the product.
- tour: tour ID
- tour_dossier: Tour dossier ID
- __rooms__: A list of available rooming options
- addons: A list of available add ons
- lowest_pp2a_prices: The lowest Adult price per person (based on double occupancy) for this departure across all rooms,
- local_payments: List of potential payment that need to be made in person
- travel_ready_policy: The Travel Ready Policy associated with this Departure
- components: a list of components associated with this departure
- structured_itineraries: A list of itineraries for this departure
- relationships: Relationships list the related departures that are available to be sold along side this departure.

### rooms

a list of rooms available for the departure

In [24]:
depart.rooms

[<DepartureRoom (Standard)>]

that are used to access details around availability, cost and promotions

In [25]:
depart.rooms[0]

<DepartureRoom (Standard)>

- code: code of room type
- name: name of room type
- flags: a list of thing that need to be flagged
- availability: details relating to availability
- price_bands: a list of price bands, one for each currency that contains amounts and promotions
- addons: available components that can be added

## relationships

In [26]:
tour_dossier.relationships[0].to_dict()

{'type': 'MINI_ADVENTURE',
 'sub_type': 'PRE_TOUR',
 'tour_dossier': {'id': '25561',
  'href': 'https://rest.gadventures.com/tour_dossiers/25561',
  'product_line': 'RAHDDN'},
 'valid_during_ranges': [{'start_date': '2021-04-01', 'end_date': None}]}

# other

# extract data for 1 departure

first departure in april 2024

In [27]:
filtered_departures = client.departures.filter(start_date="2024-04-01", )
filtered_departures.count()

130

In [28]:
my_generator = filtered_departures.all()
departures_list = []
for value in my_generator:
    departures_list.append(value)
    break

In [29]:
first_departure = departures_list[0]

In [30]:
# tour name
tour_name = first_departure.tour_dossier.name

In [31]:
# tour url
tour_url = first_departure.tour_dossier.site_links[0]['href']

In [32]:
# tour categories list
categories_id_list = [i['id'] for i in first_departure.tour_dossier.categories]
categories_id_list

['21', '30', '36', '131', '59']

In [33]:
# tour categories
tour_categories_dictionaries = {}

for category in first_departure.tour_dossier.categories:
    if category["category_type"]['label'] in tour_categories_dictionaries:
        tour_categories_dictionaries[category["category_type"]['label']] += category['name']
    else:
        tour_categories_dictionaries[category["category_type"]['label']] = category['name']
        
tour_categories_dictionaries

{'Travel Style': 'Classic',
 'Service Level': 'Standard',
 'Physical Grading': '3 - Average',
 'Merchandising': 'Planeterra Project',
 'Trip Type': 'Small Group'}

In [34]:
# tour geography
visited_countries_list = []
for country in first_departure.tour_dossier.geography['visited_countries']:
    visited_countries_list.append(country["name"])

In [35]:
# structured itinerary id
structured_itineraries_id_list = []
for i in first_departure.structured_itineraries:
    structured_itineraries_id_list.append(i.variation_id)

In [36]:
# departure start date
start_date = first_departure.start_date

In [37]:
from datetime import timedelta
# duration
duration = first_departure.finish_date - first_departure.start_date + timedelta(days=1)

In [38]:
# extract room name, price band and amount
room_name_list = []
price_bands_list = []
amount_list = []

for i in first_departure.rooms:
    data = i.to_dict()
    room_name_list.append(data["name"])
    
    for price_brand in data["price_bands"]:
        price_bands_list.append(price_brand["name"])
        
        for price in price_brand["prices"]:
            if price['currency'] == "GBP":
                amount_list.append(price['amount'])
        

In [39]:
dataframe_dictionary = {}
dataframe_dictionary['tour_name'] = [tour_name]
dataframe_dictionary['visited_countries'] = visited_countries_list
dataframe_dictionary['start_date'] = [start_date]
dataframe_dictionary['duration'] = [duration]
dataframe_dictionary['cost'] = amount_list
dataframe_dictionary['url'] = [tour_url]

In [40]:
dataframe_dictionary = dict(dataframe_dictionary, **tour_categories_dictionaries)

In [41]:
#text_data
itinerary = ''
itinerary += f"Name of the tour: {tour_name}\n"
itinerary += f'Countries visited: {"".join(visited_countries_list)}\n'
itinerary += f'Start date: {str(start_date)}\n'
itinerary += f'Trip duration: {str(duration)[:-9]}\n'
itinerary += f'Trip cost: £{"".join(amount_list)}\n'

In [42]:
for category in categories_id_list:
    category_dict = client.tour_categories.get(category).to_dict()
    itinerary += f"{category_dict['category_type']['name']}: {category_dict['name']} - {category_dict['description']}\n"

In [43]:
itinerary += 'Itinerary:\n'
itinerary_days = first_departure.structured_itineraries[0]
for day in itinerary_days.days:
    itinerary += f'Day {day.day}: {day.summary}\n'

In [44]:
print(itinerary)

Name of the tour: Iceland Northern Lights & Golden Circle
Countries visited: Iceland
Start date: 2024-04-01
Trip duration: 5 days
Trip cost: £1279.00
Travel Style: Classic - All of the highlights, culture, access, and I-can’t-believe-we-did-that moments, all at a great price.
Service Level: Standard - Comfortable tourist-class accommodations with character; mix of public and private transport.
Physical Grading: 3 - Average - Some tours may include light hiking, biking, rafting, or kayaking in addition to walking.
Merchandising: Planeterra Project - None
Trip Type: Small Group - Group trips average 12 travellers per departure, depending on the adventure. The maximum is usually no more than 16, but some can be smaller or bigger, depending on the trip. Check individual trips for details.
Itinerary:
Day 1: Arrive at any time.
Day 2: Travel from Reykjavík to explore a lava cave in the Bláfjöll Mountains. In the afternoon, get up close and personal with the famous Icelandic horse before retu

In [45]:
with open(f'{tour_name} - {" ".join(structured_itineraries_id_list)}.txt', "w") as file:
    file.write(itinerary)

In [46]:
import pandas as pd
df = pd.DataFrame(dataframe_dictionary)
df

tour_name visited_countries  start_date  \
0  Iceland Northern Lights & Golden Circle           Iceland  2024-04-01   

  duration     cost                                                url  \
0   5 days  1279.00  https://www.gadventures.com/trips/iceland-nort...   

  Travel Style Service Level Physical Grading       Merchandising    Trip Type  
0      Classic      Standard      3 - Average  Planeterra Project  Small Group

# 3 from one days data

In [47]:
filtered_departures = client.departures.filter(start_date="2024-04-19")

In [48]:
filtered_departures.count()

142

In [49]:
my_generator = filtered_departures.all()

In [50]:
departures_dictionaries_list = []
reference_count = 0
for value in my_generator:
    reference_count += 1
    if reference_count % 10 == 0:
        print(reference_count)
    try:
        if value.availability['status'] == "AVAILABLE" and value.tour_dossier != None: 
            print(value.id)
            trip_list = [category['name'] for category in value.tour_dossier.categories if category['category_type']['label'] ==  "Trip Type"]
            if 'Small Group' in trip_list:
                for i in range(len(value.structured_itineraries)):

                    # get tour name
                    tour_name = value.tour_dossier.name

                    # structured itinerary id
                    structured_itineraries_name = value.structured_itineraries[i].name

                    # get the list of countries visited
                    visited_countries_list = []
                    for country in value.tour_dossier.geography['visited_countries']:
                        visited_countries_list.append(country["name"])

                    # get start date
                    start_date = value.start_date

                    # get duration
                    duration = value.finish_date - value.start_date + timedelta(days=1)

                    # get pricing form room categories
                    room_categories_dictionaries = {}

                    for room in value.rooms:
                        data = room.to_dict()
                        for price_brand in data["price_bands"]:
                            for price in price_brand["prices"]:
                                if price['currency'] == "GBP":               
                                    if f'{data["name"]} - {price_brand["name"]}' in room_categories_dictionaries:
                                        room_categories_dictionaries[f'{data["name"]} - {price_brand["name"]}'] += price['amount']
                                    else:
                                        room_categories_dictionaries[f'{data["name"]} - {price_brand["name"]}'] = price['amount']

                    # get tour url 
                    tour_url = value.tour_dossier.site_links[0]['href']
                    
                    # get tour description
                    tour_description = value.tour_dossier.description

                    # get tour categories
                    categories_id_list = [i['id'] for i in value.tour_dossier.categories]

                    tour_categories_dictionaries = {}

                    for category in value.tour_dossier.categories:
                        if category["category_type"]['label'] in tour_categories_dictionaries:
                            tour_categories_dictionaries[category["category_type"]['label']] += ' ' + category['name']
                        else:
                            tour_categories_dictionaries[category["category_type"]['label']] = category['name']

                    # create dictionary
                    dataframe_dictionary = {}
                    dataframe_dictionary['tour_name'] = tour_name
                    dataframe_dictionary['itinerary_name'] = structured_itineraries_name
                    dataframe_dictionary['visited_countries'] = " ".join(visited_countries_list)
                    dataframe_dictionary['start_date'] = start_date
                    dataframe_dictionary['duration'] = duration
                    dataframe_dictionary['url'] = tour_url
                    dataframe_dictionary['tour_description'] = tour_description

                    # combine dictionaries
                    dataframe_dictionary = dict(dataframe_dictionary, **tour_categories_dictionaries, **room_categories_dictionaries)

                    # add full dict to list
                    departures_dictionaries_list.append(dataframe_dictionary)

                    # text file
                    itinerary = ''
                    itinerary += f"Name of the tour: {tour_name}\n"
                    itinerary += f'Countries visited: {" ".join(visited_countries_list)}\n'
                    itinerary += f'Trip duration: {str(duration)[:-9]}\n'

#                     for key, description in room_categories_dictionaries.items():
#                         itinerary += f'{key}: {description}\n'

                    for category in categories_id_list:
                        category_dict = client.tour_categories.get(category).to_dict()
                        itinerary += f"{category_dict['category_type']['name']}: {category_dict['name']} - {category_dict['description']}\n"


                    itinerary += 'Itinerary:\n'
                    itinerary_days = value.structured_itineraries[i]
                    for day in itinerary_days.days:
                        itinerary += f'Day {day.day}: {day.summary}\n'


                    with open(f'raw_data/itinerary_text/{tour_name} {structured_itineraries_name}.txt', "w") as file:
                        file.write(itinerary)

                    print(f'{tour_name} - {structured_itineraries_name} - {value.tour_dossier.id} - {len(value.structured_itineraries)}')
        
    except:
        print("403 issue")


        # ID: {value.tour_dossier.id}. Number of itineraries: {len(value.structured_itineraries)}

1196464
403 issue
1198260
Laos: Sunrises & Street Food -  - 24821 - 1
10
1200164
Laos to Thailand: Night Markets & Mekong Cruising -  - 24823 - 1
1200263
Laos to Northern Thailand: Treks & Trails -  - 25030 - 1
20
1206493
30
1207235
1207601
1208333
403 issue
403 issue
1209431
1209797
403 issue
1210529
40
1225257
Everest Base Camp Trek -  - 22946 - 1
1225886
Golden Triangle -  - 24213 - 1
50
1229178
Maldives Dhoni Cruise -  - 21203 - 1
1229376
Classic Bali -  - 22791 - 1
1229560
Borneo – East Sabah Adventure -  - 24511 - 1
1229745
Discover Japan -  - 22800 - 1
1229896
Japan Express: Osaka to Tokyo -  - 22801 - 1
1230211
Best of Cambodia & Northern Thailand -  - 25018 - 1
1231033
Best of South Korea -  - 24992 - 1
60
1232424
Highlights of Jordan -  - 22862 - 1
1232429
Egypt & Jordan Adventure -  - 22854 - 1
1232430
Best of Egypt -  - 22853 - 1
1232875
Jordan Multisport -  - 23286 - 1
1233035
Highlights of Morocco -  - 22844 - 1
1233345
Coastal Morocco: Waves & Market Stalls -  - 25021 - 

In [51]:
df = pd.DataFrame(departures_dictionaries_list)
df.to_csv('raw_data/one_day_test.csv', index=False)
df

tour_name itinerary_name  \
0                        Laos: Sunrises & Street Food                  
1   Laos to Thailand: Night Markets & Mekong Cruising                  
2           Laos to Northern Thailand: Treks & Trails                  
3                              Everest Base Camp Trek                  
4                                     Golden Triangle                  
5                               Maldives Dhoni Cruise                  
6                                        Classic Bali                  
7                       Borneo – East Sabah Adventure                  
8                                      Discover Japan                  
9                       Japan Express: Osaka to Tokyo                  
10               Best of Cambodia & Northern Thailand                  
11                                Best of South Korea                  
12                               Highlights of Jordan                  
13                           Egypt & Jordan Adventure                  
14                                      Best of Egypt                  
15                                  Jordan Multisport                  
16                              Highlights of Morocco                  
17             Coastal Morocco: Waves & Market Stalls                  
18                                    Southern Divide                  
19            The Scenic Route - Lima to Buenos Aires                  
20                                Amazon to the Andes     Inca Trail   
21                                Amazon to the Andes     Lares Trek   
22                                Amazon to the Andes     Cusco Stay   
23                                Inca Discovery Plus     Inca Trail   
24                                Inca Discovery Plus     Lares Trek   
25                                Inca Discovery Plus     Cusco Stay   
26                                     Inca Discovery     Inca Trail   
27                                     Inca Discovery     Lares Trek   
28                                     Inca Discovery     Cusco Stay   
29                                     The Inca Trail     Inca Trail   
30                                     The Inca Trail     Lares Trek   
31                                     The Inca Trail     Cusco Stay   
32                                     The Lares Trek     Lares Trek   
33                                     The Lares Trek     Cusco Stay   
34                     Trekking the Inca Trail: 5D/4N     Inca Trail   
35                     Trekking the Inca Trail: 4D/3N     Inca Trail   
36             Lima to La Paz: Sandboarding & Sunsets     Inca Trail   
37             Lima to La Paz: Sandboarding & Sunsets     Lares Trek   
38             Lima to La Paz: Sandboarding & Sunsets     Cusco Stay   
39            Lima to Buenos Aires: Deserts & Dancing     Inca Trail   
40            Lima to Buenos Aires: Deserts & Dancing     Lares Trek   
41            Lima to Buenos Aires: Deserts & Dancing     Cusco Stay   
42                     Camino de Santiago Encompassed                  
43                 Spain & Portugal: Flamenco & Tapas                  
44  Spain, Portugal & Morocco: Tapas, Medinas & Su...                  
45                Best of Uzbekistan and Turkmenistan                  
46           Ecuador: Amazon, Hot Springs & Volcanoes                  
47  Mexico Coast-to-Coast: Puerto Vallarta to the ...                  
48                           Essential Central Mexico                  
49  The Many Sides of Mexico: Puerto Vallarta to O...                  
50                        Lima to Rio: Coast to Coast     Inca Trail   
51                        Lima to Rio: Coast to Coast     Lares Trek   
52                        Lima to Rio: Coast to Coast     Cusco Stay   
53                                Cambodia Experience                  
54     Vietnam: Historic Cities & Halong Bay Cruising                  
55          

In [54]:
found_itineraries = ['Best of Spain & Portugal', 'Highlights of Portugal']

In [65]:
summary = ""
for itinerary in found_itineraries:
    filtered_df = df[df['tour_name'] == itinerary]
    summary += f'Itinerary: {filtered_df["tour_name"].values[0]}\n'
    summary += f'Tour description: {filtered_df["tour_description"].values[0]}\n\n'
print(summary)

Itinerary: Best of Spain & Portugal
tour description: Dynamic culture, compelling history, to-die-for cuisine — the Iberian Peninsula really is a true example of riches for every sense and interest. On this adventure, you’ll soak up the Moorish history of Lisbon and Cordoba, be treated to a lively flamenco performance in Seville, explore the astounding red-clay fortress of the Alhambra, and witness Gaudí’s Barcelona with your own eyes. If that sounds like a lot, be prepared; we’ve only just begun.

Itinerary: Highlights of Portugal
tour description: In Portugal, the sun is bright, the food is delicious, and the living is easy. Wander the hilly streets of coastal Lisbon, where you can munch on a custard tart while gazing at sun-dappled architecture. Get your fill of history and culture at UNESCO World Heritage Sites like the Monastery of Batalha, Coimbra University, and the picturesque town of Sintra (optional). End the tour in Porto, where the majestic Douro Valley begs for you to loun

In [64]:
filtered_df = df[df['tour_name'] == 'Best of Spain & Portugal']
filtered_df['visited_countries'].values[0]

'Spain Portugal'